core/src/main/scala/org/apache/spark/api/r/RBackendHandler.scala

```diff
@@ -263,18 +264,19 @@ private[r] object JVMObjectTracker {
 
   // TODO: This map should be thread-safe if we want to support multiple
   // connections at the same time
-  private[this] val objMap = new HashMap[String, Object]
+  private[this] val objMap: ConcurrentMap[String, Object] =
+    new MapMaker().weakValues().makeMap[String, Object]()
 
   // TODO: We support only one connection now, so an integer is fine.
   // Investigate using use atomic integer in the future.
   private[this] var objCounter: Int = 0
 
   def getObject(id: String): Object = {
-    objMap(id)
+    objMap.get(id)
   }
 
   def get(id: String): Option[Object] = {
-    objMap.get(id)
+    Option(objMap.get(id))
   }
 
   def put(obj: Object): String = {
```

Comment:
Agree, this can just be changed a little more to also handle that issue. They end up being interrelated here.